orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -511,6 +511,7 @@ void addObjectMethodsForTraining(py::module& m) {
       })
       .def("run_backward", [](TrainingAgent* agent, const std::vector<OrtValue>& feeds, std::vector<OrtValue>& fetches, PartialGraphExecutionState* state) -> void {
         Status status = agent->RunBackward(feeds, fetches, *state);
+        state->~PartialGraphExecutionState();
```

Comment:
Mentioned this offline, but I don't think explicitly invoking the destructor is the right thing to do here. The options I can see are:

1. Remove this line, and let the `std::unique_ptr<PartialGraphExecutionState>`  (in the pybind wrapping class) destructor take care of it for you. (But perhaps this happens too late or too unpredictably?)
2. Find some way to pass the boxed `std::unique_ptr<PartialGraphExecutionState>` to this method, so that it can "consume" the state by calling `reset()` on it.

Reply:
Suggestion 2 would be the "good sink" pattern from https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/, which I think most closely matches what you're trying to do here.

Reply:
Would this work for megatron?

Reply:
Turns out python has a del keyword that deferments the reference count and if that reaches zero then CPython immediately frees the memory. Since the object is created on the python end it seems natural to also try to memory associated with it also on the python side.

Reply:
@KeDengMS the latest commit in this PR should work. @pengwa can verify this commit on Megatron.